src/platform/amd/acp_7_0/lib/clk.c

```diff
@@ -0,0 +1,491 @@
+
+void audio_pll_power_off(void);
+void audio_pll_power_on(void);
+void clk_dfs_bypassexit(void);
+void audio_pll_mode_switch(uint32_t mode, uint32_t fcw_int, uint32_t fcw_frac, uint32_t fcw_denom,
+			   uint32_t pllspinediv);
+void clk7_init_vco(void);
+void acp_7_0_reg_wait(void);
+void acp_7_0_get_boot_ref_clock(float *boot_ref_clk);
+
+typedef enum  {
+	PLL_MODE_100MHZ_NORMAL,
+	PLL_MODE_48MHZ_NORMAL,
+	PLL_MODE_32KHZ_LPPM,
+	PLL_MODE_48MHZ_LPPM,
+	PLL_MODE_100MHZ_LPPM
+} PLL_MODE;
+
+/* Enumeration for the Clock Types */
+typedef enum _acp_clock_type_ {
+	acp_aclk_clock,
+	acp_sclk_clock,
+	acp_clock_type_max,
+	acp_clock_type_force = 0xFF
+} acp_clock_type_t;
+
+static int acp_reg_read_via_smn(uint32_t reg_offset,
+				uint32_t size)
+{
+	uint32_t reg_value;
+	uint32_t delay_cnt = 10000;
+	uint32_t smn_client_base_addr = (reg_offset >> 10);
+	uint32_t region_start_addr = (smn_client_base_addr << 10);
+	uint32_t apertureid = ((reg_offset >> 20) & 0xFFF);
+	acp_srbm_cycle_sts_t  acp_srbm_cycle_sts;
+
+	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_CONFIG), apertureid);
+	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_BASE_ADDR), smn_client_base_addr);
+	reg_value = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+	(ACP_MASTER_REG_ACCESS_ADDRESS + reg_offset - region_start_addr + ACP_FIRST_REG_OFFSET));
+	if (reg_value)
+		reg_value = 0;
+	acp_srbm_cycle_sts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE +
+							       ACP_SRBM_CYCLE_STS);
+	while (delay_cnt > 0) {
+		if (!acp_srbm_cycle_sts.bits.srbm_clients_sts)
+			return (uint32_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CLIENT_RDDATA);
+		acp_srbm_cycle_sts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE +
+									ACP_SRBM_CYCLE_STS);
+		delay_cnt--;
+	}
+	return -1;
+}
+
+static void acp_reg_write_via_smn(uint32_t reg_offset,
+				  uint32_t value, uint32_t size)
+{
+	uint32_t	delay_cnt = 10000;
+	uint32_t        smn_client_base_addr    = (reg_offset >> 10);
+	uint32_t        region_start_addr = (smn_client_base_addr << 10);
+	uint32_t	apertureid = ((reg_offset >> 20) & 0xFFF);
+	acp_srbm_cycle_sts_t      acp_srbm_cycle_sts;
+
+	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_CONFIG), apertureid);
+	io_reg_write((PU_REGISTER_BASE + ACP_SRBM_CLIENT_BASE_ADDR), smn_client_base_addr);
+	io_reg_write((PU_REGISTER_BASE +
+		(ACP_MASTER_REG_ACCESS_ADDRESS + reg_offset - region_start_addr +
+		 ACP_FIRST_REG_OFFSET)),
+		value);
+	acp_srbm_cycle_sts =
+		(acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE + ACP_SRBM_CYCLE_STS);
+	while (delay_cnt > 0) {
+		acp_srbm_cycle_sts = (acp_srbm_cycle_sts_t)io_reg_read(PU_REGISTER_BASE +
+								ACP_SRBM_CYCLE_STS);
+		if (!acp_srbm_cycle_sts.bits.srbm_clients_sts)
+			return;
+		delay_cnt--;
+	}
+}
+
+void acp_7_0_reg_wait(void)
+{
+	int test_count = 0;
+	int val = 0;
+
+	for (test_count = 0; test_count < 255; test_count++) {
+		val = acp_reg_read_via_smn(CLK7_CLK_FSM_STATUS, sizeof(int));
+		val = val * 0;
```

Comment:
Actually `val` variable is being used to read the reg `CLK7_CLK_FSM_STATUS`, reading is being done to create some delay.  `val = val * 0;` is to avoid the error "variable declared, but not used".

Reply:
Then you don't need `val`
```C
// Wait for something / add a delay 
(void) acp_reg_read_via_smn(CLK7_CLK_FSM_STATUS, sizeof(int))
```